### PR TITLE
H373 — Transolver++ Local Adaptive Slice Pooling (k=16 kNN, temp=0.1)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Replace global slice softmax pooling in Transolver with local-neighborhood-constrained pooling (k=16 nearest neighbors, temperature=0.1).**

The current Transolver slice pool assigns each surface point to one of K=32 slices via a GLOBAL softmax — meaning a single slice may average over physically disconnected regions (hood + underbody, A-pillar + diffuser). For τ_z, the relevant physics is **highly local** (separation lines, A-pillar vortex attachment, wake shoulder), and distant surface regions have different τ_z signatures. Global pooling forces the model to average over physically incoherent neighborhoods — degrading the fine-grained boundary-layer structure where τ_z errors are largest.

**Local adaptive pooling** restricts each slice to a k-NN neighborhood: surface point i can contribute only to slices whose centroid is among its k=16 nearest neighbors (Euclidean on x,y,z). The softmax temperature is τ=0.1 to sharpen the assignment. Zero new parameters — the projection matrices are unchanged, only the pooling INDEX changes.

**Why this is fundamentally different from prior nulls:**
- Loss reweighting axis (7 nulls H338-H368): scalar reweighting can't move WSS_z. **This changes the model's forward-pass attention TOPOLOGY, not the loss signal.**
- Input-feature axis (4 nulls H348/H359/H360/H369): zero-pad warm-start of new geometric features failed to propagate signal through cosine-tail. **This adds NO new input features — it changes how existing features are aggregated.**
- Encoder/normal axis (4 nulls H351/H357/H358/H367): tangent-basis output heads, GeoTransolver content embedding — all closed. **This is upstream of those: the slice-pooling step that FEEDS the encoder.**
- ISAB operator family (closed at single-layer bound H370/H371): inducing-point compression breaks token-count bijection. **This preserves token bijection 100% — slice → point is unchanged, only the slice-pool routing changes.**

**Empirical grounding:** Transolver++ (Wu et al., ICML 2025, arxiv:2502.02414) shows local adaptive pooling delivers **+8-12% relative L2 on NS-2D and Elasticity** — closest-domain results to surface CFD. GAOT (ETH 2025, arxiv:2505.18781) confirms geometry-aware multiscale local pooling beats global pooling on PDE operator transformers.

## Instructions

**Branch:** `frieren/h373-transolver-pp-local-slice-pool` (created)

**Step 1 — Find and modify `target/model/transolver.py` `SlicePool.forward()` (or equivalent slice-pooling class):**

Add CLI flags:
```
--slice-pool-mode local_adaptive   # NEW: switches from global softmax → local kNN softmax
--slice-pool-neighbors 16          # k for kNN
--slice-pool-temperature 0.1       # softmax temperature inside local pool
```

Implementation approach:
- Current forward (assumed): `slice_assignment = softmax(point_features @ slice_centroid^T / temp)` — shape (N_points, K_slices), global softmax over K slices.
- New forward (local_adaptive): for each point i, compute the top-k=16 nearest slice CENTROIDS by Euclidean distance on (x,y,z); mask out all other slices to −inf before softmax. Apply softmax with temp=0.1 over only the top-k entries. Output remains shape (N_points, K_slices) but sparse with at most 16 non-zero entries per row.
- The slice-centroid positions: if learnable parameters, use the current learned values from EP13 EMA warm-start. If derived from feature space, use the current projection.
- **CRITICAL — token bijection preservation:** N_points input → K_slices intermediate → N_points output is preserved. No information compression bottleneck. The change is in the SLICE → POINT routing only.

**Step 2 — Smoke check (2-epoch debug, no warm-start, batch_size=4):**
```bash
torchrun --standalone --nproc-per-node=8 train.py \
  --epochs 2 --debug \
  --slice-pool-mode local_adaptive --slice-pool-neighbors 16 --slice-pool-temperature 0.1 \
  --wandb-group h373-frieren-smoke \
  --wandb-name "frieren/h373-smoke-local-slice"
```
Verify: no NaN in slice_assignment matrix, training loss decreases vs random init.

**Step 3 — Phase 1 (EP13 → EP14-EP16, 8 GPUs DDP, 3-epoch cosine tail):**

If smoke clean, launch primary:
```bash
H185_EP13_EMA="outputs/ensemble_cache/run-yw2a5dyl-epoch-13-ema/checkpoint.pt"

torchrun --standalone --nproc-per-node=8 train.py \
  --warm-start-from "$H185_EP13_EMA" \
  --epochs 3 --start-epoch 13 \
  --optimizer lion --lr 9e-5 --lr-cosine-t-max 3 --lr-min 1e-6 \
  --weight-decay 5e-4 --batch-size 4 \
  --tau-y-loss-weight 1.3 --tau-z-loss-weight 1.67 \
  --surface-loss-weight 2.0 --volume-loss-weight 0.5 \
  --grad-clip-norm 0.5 --ema-decay 0.999 \
  --use-qk-norm --rff-num-features 16 --pos-encoding-mode string_separable \
  --model-layers 5 --hidden-dim 512 --heads 4 --slices 128 \
  --slice-pool-mode local_adaptive --slice-pool-neighbors 16 --slice-pool-temperature 0.1 \
  --vol-points-schedule "0:65536" \
  --wandb-group h373-frieren-transolver-pp \
  --wandb-name "frieren/h373-phase1-local-slice"
```

**Pre-committed kill rules:**
- EP14 val_primary > 6.05% → CLOSE (mechanism not propagating through EP13 basin)
- EP14 val_WSS_z > H336 EP13 baseline (~9.06%) → CLOSE (local pooling disrupted boundary-layer rep)
- EP14 val_primary ≤ 6.05% AND val_WSS_z drops ≥30bp → continue EP15-EP16, run TTA cascade

**Step 4 — TTA cascade (if Phase 1 passes EP14 gate):**

Apply H336 recipe to best-val checkpoint:
```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint <best_ckpt.pt> \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 --weight-noise-passes 5 --antithetic-noise \
  --weight-noise-dist student_t --weight-noise-df 4 \
  --test-time-calibration \
  --wandb-group h373-frieren-tta \
  --wandb-name "frieren/h373-tta-cal"
```
Then test split with same recipe.

## Baseline

**Current SOTA H342 (PR #1526, MERGED):**
- val_abupt_calibrated = **5.8962%**
- test_abupt_calibrated = **5.7357%**
- test_VP = 3.3751%, test_SP = 3.6124%, test_WSS = 6.6351%, **test_WSS_z = 8.6122%**

**Merge gate (strict AND):** val_cal < 5.8962% AND test_cal < 5.7357%

**Estimated wall-clock:** ~3.5h/epoch × 4 epochs = ~14h on 8 GPUs (Phase 1 requires attention kernel rewrite, may take longer). TTA cascade ~7h. Total ≤ 21h.

**Implementation references:**
- Transolver++ (Wu et al., ICML 2025) arxiv:2502.02414 — primary reference; their open-source code at github.com/thuml/Transolver++ has `LocalAdaptiveSlicePool` class
- Transolver (Wu et al., ICML 2024) arxiv:2402.02366 — our baseline architecture
- PointConv (Wu et al., CVPR 2019) — local kNN pooling on point clouds as precedent

**Constraints:**
- DDP×8 mandatory
- τ_z normalization LOCKED at 1.67 (never modify)
- Read-only: `data/loader.py`, `data/preload.py`, `data/split_manifest.json`
- Zero new parameters (only INDEX change inside SlicePool)
- Preserve token-count bijection (N → K → N, NEVER compress)

Use `--wandb_group h373-frieren-transolver-pp`. Submit terminal `SENPAI-RESULT` marker on completion.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT frieren: Read PR #1568. Implementation confirmed already complete in `model.py` (lines 313–344: `TransolverAttention.create_slices()` with top-k masked softmax, k=16, τ=0.1) and `train.py` (CLI flags `--slice-pool-mode`, `--slice-pool-neighbors`, `--slice-pool-temperature` wired through Config → build_model() → SurfaceTransolver → Transformer → TransformerBlock → TransolverAttention). Warm-start checkpoint found at `outputs/drivaerml/resume_cache/yw2a5dyl/epoch-13/checkpoint.pt` (note: PR specifies `outputs/ensemble_cache/run-yw2a5dyl-epoch-13-ema/checkpoint.pt` — using actual found path). Launching smoke check now (2-epoch debug, no warm-start).

**Comment by morganmcg1** [comment]
STUDENT frieren:

## H373 Phase 1 launched

**Implementation summary** (see commit 8d4fe4c on this branch):

- `model.py` — `TransolverAttention.create_slices()` now branches on `slice_pool_mode`:
  - `global` (default): unchanged behaviour. `slice_logits = in_project_slice(x_mid) / self.temperature; softmax`.
  - `local_adaptive`: top-k masked softmax. `topk_idx = slice_logits_raw.topk(k, dim=-1)`, build a bool keep-mask, then `clamp(-50,50) / slice_pool_temperature`, masked_fill(~keep, -inf), softmax. Zero new parameters — top-k is over the existing learned `in_project_slice` projection. `self.temperature` is frozen (`requires_grad=False`) in local mode so DDP and EMA both skip it cleanly.
  - Belt-and-suspenders guards: clamp(-50,50) before /τ stops bf16 overflow at τ=0.1 (±50 raw → ±500 scaled, safe). `nan_to_num` on the softmax output covers a degenerate all-masked row (cannot happen with k≥1 and finite logits, but kept as a silent-crash guard).
  - `slice_pool_neighbors` is capped at `num_slices` to keep top-k valid.
- `train.py` — three new Config fields + help text (`slice_pool_mode`/`slice_pool_neighbors`/`slice_pool_temperature`) wired through `build_model()`.
- `eval_tta_h252.py` — mirrored EvalConfig fields so H373 checkpoints are loadable by the TTA cascade.

**Interpretation note on "Euclidean kNN":** the PR body describes top-k as "k nearest slices in feature space" via Euclidean distance on `(x,y,z)`. With the zero-new-params constraint there are no learnable slice centroids, so I interpret "kNN" as feature-space top-k over `in_project_slice(x_mid)`. This is equivalent to picking the k slices whose learned query vectors have the largest inner product with the per-point projected feature — the same softmax-input geometry the global pool already uses, just sparsified.

**Warm-start verification (smoke):**
- 2-epoch debug run (8 GPUs, 4 cases/rank) ran end-to-end clean — loss 0.92 → 0.08, no NaN gradients, no DDP unused-param errors.
- EP13 EMA checkpoint (`yw2a5dyl/epoch-13`) loads into local_adaptive model with 0 missing / 0 unexpected keys. Sparsity check confirmed: exactly k=16 non-zero entries per (batch, head, point, slice) row.

**Phase 1 command** (running now, PID 959303 on the pod, log `logs/h373_phase1.log`):
```
torchrun --standalone --nproc-per-node=8 train.py \
  --resume-from-wandb yw2a5dyl --resume-alias epoch-13 --epochs-already-done 0 \
  --data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --agent frieren --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.3 --tau-z-loss-weight 1.67 \
  --surface-loss-weight 2.0 --volume-loss-weight 0.5 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 0 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 3 --lr-min 1e-6 --epochs 3 \
  --vol-points-schedule "0:65536" --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn --mirror-augmentation --drop-path-max 0.1 \
  --slice-pool-mode local_adaptive --slice-pool-neighbors 16 --slice-pool-temperature 0.1 \
  --save-every-epoch \
  --wandb-group h373-frieren-transolver-pp \
  --wandb-name "frieren/h373-phase1-local-slice"
```

**Flag deviation from PR body:**
- The PR specified `--warm-start-from <path> --start-epoch 13`; those flags don't exist in the current `train.py`. The equivalent canonical wiring is `--resume-from-wandb yw2a5dyl --resume-alias epoch-13 --epochs-already-done 0`, which loads the EP13 EMA artifact and starts the LR/cosine from a fresh epoch counter so `--lr-cosine-t-max 3` actually sweeps 9e-5 → 1e-6 over the 3 fine-tune epochs (instead of being short-circuited to lr_min by an advanced scheduler step). Please flag if you'd prefer a different interpretation of the schedule.
- Throughput on the pod is ~1.15 s/it × 2720 steps ≈ **52 min/epoch**, so EP1 (= advisor's "EP14") val should complete in ~70 min from launch. Kill gate: val_primary > 6.05%.

**Master tracking:**
- W&B group: `h373-frieren-transolver-pp`
- Run name: `frieren/h373-phase1-local-slice` (rank0 + 7 worker runs)
- Will post EP1 results comment as soon as the first validation completes; if it clears the 6.05% kill gate I'll let the full 3-epoch tail run and then post a complete Results comment with Phase 2 TTA numbers.

**Comment by morganmcg1** [comment]
ADVISOR HEARTBEAT — H373 Phase 1 overdue (2026-06-02 ~17:10Z)

Phase 1 was launched at 13:52Z with ~50 min/epoch. EP1 val was expected around 14:52-15:30Z — that is now **3+ hours overdue** with no update.

**Required immediate action:** Report EP1 val result or pod status.

- Is W&B run `frieren/h373-phase1-local-slice` (group `h373-frieren-transolver-pp`) still logging? What is the current step/epoch?
- If training crashed or stalled: post the last 30 lines of logs and the error.
- If EP1 completed but val wasn't posted: post `val_abupt_rel_l2_pct` now.

If the run has diverged or is producing NaN loss, stop immediately and report. The local_adaptive top-k softmax can cause sparse gradients or numerical instability — check the BF16 overflow guard (clamp(-50,50)) is active and that at least 1 token is guaranteed to be non-masked per head.

Post an update within 30 minutes.

**Comment by morganmcg1** [comment]
ADVISOR: H373 Phase 1 complete — REGRESSION, closing experiment.

**Phase 1 results (EP3 best, W&B run `hp9qxlp4`):**

| Metric | Value |
|--------|-------|
| val_RAW (abupt_axis_mean_rel_l2_pct) | **6.1276%** |
| val_WSS_z | 9.303% |
| val_WSS | 6.905% |
| val_pressure | 4.063% |
| best_epoch | 3 |
| source | EMA |

**SOTA cluster for reference:**
- H347 Arm B: 5.9224%
- H359: 5.9245%
- H365: 5.9273%
- H342 baseline EP13: ~5.93%

**Assessment:** The Local Adaptive Slice Pooling intervention (top-k sparsified softmax, k=16, τ=0.1 in `create_slices()`) produced a clear **+20bp regression** vs the SOTA cluster. val_RAW of 6.1276% is worse than EP13 EMA baseline (~5.93%). Even the improvement from EP1→EP3 did not recover toward baseline.

**Finding Z (25th axis closed): Top-k sparse slice attention hurts.** Replacing Transolver's uniform-softmax slice assignment with top-k sparsified attention degrades val_RAW by ~+20bp. The original uniform softmax appears better for this geometry. This is consistent with our broader pattern that encoder representation is the limit — but sparse slice pooling is the wrong axis.

**Decision: CLOSE.** Phase 2/TTA not warranted — a 20bp regression cannot be recovered by calibration. Closing this PR.

Thank you frieren — the implementation was clean and the experiment was properly executed. The null result (finding that sparse slice attention hurts) is a useful constraint.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1568 because H373 CLOSED — Finding Z (25th axis): top-k sparse slice attention (k=16, τ=0.1) in Transolver create_slices() degrades val_RAW by ~+20bp (6.1276% vs SOTA cluster 5.9224-5.9273%). Phase 2/TTA not warranted. Encoder representation limit confirmed again from a new angle — sparse pooling is the wrong intervention..
